### PR TITLE
BROOKLYN-381: support nested DSL for $brooklyn:entity()

### PR DESCRIPTION
**Pull request by aledsage**
Note this builds on https://github.com/apache/brooklyn-server/pull/416; please review + merge that PR first. I'll then rebase so the initial commits disappear.

@googlielmo you'll be interested in this.

### CODE REVIEW

**Comment by tbouron** [comment]
I'm not really familiar with the code but LGTM. Unit test pass and I tested it with https://github.com/brooklyncentral/brooklyn-riak-cluster/pull/1

**Comment by googlielmo** [comment]
Paired with @aledsage when doing this. LGTM.

**Comment by aledsage** [comment]
Thanks @tbouron @googlielmo - merging now.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java [hunk 1]
```diff
@@ -182,63 +257,112 @@ public Entity call() throws Exception {
...
+            String desiredComponentId;
+            if (componentId == null) {
+                if (componentIdSupplier == null) {
+                    throw new IllegalArgumentException("No component-id or component-id supplier, when resolving entity in scope '" + scope + "' wrt " + entity);
+                }
+                
+                Maybe<Object> maybeComponentId = Tasks.resolving(componentIdSupplier)
+                        .as(Object.class)
+                        .context(getExecutionContext())
+                        .immediately(immediate)
```
Comment (neykov) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java]: For the `false` case would make more sense to skip `if (timeout == null) timeout = ValueResolver.NON_BLOCKING_WAIT;` in the method.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java [hunk 2]
```diff
@@ -182,63 +257,112 @@ public Entity call() throws Exception {
...
+                        .description("Resolving component-id from " + componentIdSupplier)
+                        .getMaybe();
+                
+                if (immediate) {
+                    if (maybeComponentId.isAbsent()) {
+                        return Maybe.absent(Maybe.getException(maybeComponentId));
+                    }
+                }
+                
+                // Support being passes an explicit entity via the DSL
```
Comment (neykov) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java]: typo `passes`

**Comment by aledsage** [comment]
@neykov sorry, I missed your comments before merging! I'll look through those, and address them in a subsequent pull request.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java [hunk 3]
```diff
@@ -182,63 +257,112 @@ public Entity call() throws Exception {
...
                 (scope==Scope.GLOBAL ? "" : ", in scope "+scope+" wrt "+entity+
                 (scopeComponent!=null ? " ("+scopeComponent+" from "+entity()+")" : "")));
-        }        
+        }
+        
+        private ExecutionContext getExecutionContext() {
+            EntityInternal contextEntity = (EntityInternal) BrooklynTaskTags.getTargetOrContextEntity(Tasks.current());
+            ExecutionContext execContext =
+                    (contextEntity != null) ? contextEntity.getExecutionContext()
+                                            : BasicExecutionContext.getCurrentExecutionContext();
```
Comment (neykov) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java]: Should never be the case, better throw.

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java [hunk 4]
```diff
@@ -182,63 +257,112 @@ public Entity call() throws Exception {
...
             // TODO may want to block and repeat on new entities joining?
-            throw new NoSuchElementException("No entity matching id " + componentId+
+            throw new NoSuchElementException("No entity matching id " + desiredComponentId+
                 (scope==Scope.GLOBAL ? "" : ", in scope "+scope+" wrt "+entity+
                 (scopeComponent!=null ? " ("+scopeComponent+" from "+entity()+")" : "")));
-        }        
+        }
+        
+        private ExecutionContext getExecutionContext() {
+            EntityInternal contextEntity = (EntityInternal) BrooklynTaskTags.getTargetOrContextEntity(Tasks.current());
```
Comment (neykov) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java]: Already available from `entity()`

File: camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java (hunk 1 above)
Comment (neykov) [inline camp/camp-brooklyn/src/main/java/org/apache/brooklyn/camp/brooklyn/spi/dsl/methods/DslComponent.java]: I see now you already changed the implementation.

**Comment by neykov** [comment]
LGTM, minor comments only.

**Pull request merged**